arc20.sol

```diff
@@ -0,0 +1,116 @@
+        assetID = _assetID;
+    }
+    
+    function getBalance(address addr, uint256 aID) public returns (uint256) {
+        (bool success, bytes memory data) = balanceAddr.call(abi.encodePacked(addr, aID));
+        require(success, "getBalance failed");
+        return abi.decode(data, (uint256));
+    }
+    
+    function assetCall(address addr, uint256 aID, uint256 assetAmount, bytes memory callData) public returns (bytes memory) {
+        (bool success, bytes memory data) = transferAddr.call(abi.encodePacked(addr, aID, assetAmount, callData));
+        require(success, "assetCall failed");
+        return data;
+    }
+    
+    function contractBalance() internal returns (uint256) {
+        return getBalance(address(this), assetID);
+    }
+    
+    function updateContractBalance() public {
+        _totalSupply = contractBalance();
+    }
+    
+    event Deposit(address indexed depositor, uint256 value);
+    event Withdrawal(address indexed withdrawer, uint256 value);
+    
+    function deposit() public returns (uint256) {
+        uint256 balance = contractBalance();
+        require(balance >= _totalSupply, "Balance was less than total supply. Critical error occurred.");
+        uint256 depositAmount = balance - _totalSupply;
+        _totalSupply = balance;
+        balanceOf[msg.sender] += depositAmount;
+        emit Deposit(msg.sender, depositAmount);
+        return depositAmount;
+    }
+    
+    function withdraw(uint256 amount) public {
+        require(balanceOf[msg.sender] >= amount, "Insufficient funds for withdrawal");
+        
+        // use multicoin call here to transfer
+        balanceOf[msg.sender] -= amount;
+        _totalSupply -= amount;
+        // Transfer multicoin asset to sender
+        assetCall(msg.sender, assetID, amount, "");
+        // TODO remove this check, must allow balances to be different otherwise
+        // someone sending funds without calling deposit could cause withdraw to
+        // malfunction
+        uint256 newBalance = contractBalance();
+        require(_totalSupply == newBalance, "Total supply does not equal contract balance");
+        emit Withdrawal(msg.sender, amount);
+    }
+    
+    event Transfer(address indexed from, address indexed to, uint256 value);
+
+    function transfer(address to, uint256 value) public returns (bool success) {
+        require(balanceOf[msg.sender] >= value, "insufficient balance for transfer");
+
+        balanceOf[msg.sender] -= value;  // deduct from sender's balance
+        balanceOf[to] += value;          // add to recipient's balance
+        emit Transfer(msg.sender, to, value);
+        return true;
+    }
+
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+
+    mapping(address => mapping(address => uint256)) public allowance;
+
+    function approve(address spender, uint256 value)
+        public
+        returns (bool success)
+    {
+        allowance[msg.sender][spender] = value;
```

Comment:
This seems incredibly unsafe, but apparently that's the erc20 interface... so I guess it's fine.